src/blas/backends/cublas/cublas_scope_handle.cpp

```diff
@@ -35,106 +30,46 @@ namespace cublas {
  * takes place if no other element in the container has a key equivalent to
  * the one being emplaced (keys in a map container are unique).
  */
-#ifdef ONEMKL_PI_INTERFACE_REMOVED
-thread_local cublas_handle<ur_context_handle_t> CublasScopedContextHandler::handle_helper =
-    cublas_handle<ur_context_handle_t>{};
-#else
-thread_local cublas_handle<pi_context> CublasScopedContextHandler::handle_helper =
-    cublas_handle<pi_context>{};
-#endif
+thread_local std::shared_ptr<std::unordered_map<CUdevice, cublasHandle_t>>
+    CublasScopedContextHandler::device_handle_map{ nullptr };
 
 CublasScopedContextHandler::CublasScopedContextHandler(sycl::queue queue, sycl::interop_handle& ih)
-        : ih(ih),
-          needToRecover_(false) {
-    placedContext_ = new sycl::context(queue.get_context());
-    auto cudaDevice = ih.get_native_device<sycl::backend::ext_oneapi_cuda>();
-    CUresult err;
-    CUcontext desired;
-    CUDA_ERROR_FUNC(cuCtxGetCurrent, err, &original_);
-    CUDA_ERROR_FUNC(cuDevicePrimaryCtxRetain, err, &desired, cudaDevice);
-    if (original_ != desired) {
-        // Sets the desired context as the active one for the thread
-        CUDA_ERROR_FUNC(cuCtxSetCurrent, err, desired);
-        // No context is installed and the suggested context is primary
-        // This is the most common case. We can activate the context in the
-        // thread and leave it there until all the PI context referring to the
-        // same underlying CUDA primary context are destroyed. This emulates
-        // the behaviour of the CUDA runtime api, and avoids costly context
-        // switches. No action is required on this side of the if.
-        needToRecover_ = !(original_ == nullptr);
-    }
-}
-
-CublasScopedContextHandler::~CublasScopedContextHandler() noexcept(false) {
-    if (needToRecover_) {
-        CUresult err;
-        CUDA_ERROR_FUNC(cuCtxSetCurrent, err, original_);
-    }
-    delete placedContext_;
-}
-
-void ContextCallback(void* userData) {
-    auto* ptr = static_cast<std::atomic<cublasHandle_t>*>(userData);
-    if (!ptr) {
-        return;
-    }
-    auto handle = ptr->exchange(nullptr);
-    if (handle != nullptr) {
-        cublasStatus_t err1;
-        CUBLAS_ERROR_FUNC(cublasDestroy, err1, handle);
-        handle = nullptr;
-    }
-    else {
-        // if the handle is nullptr it means the handle was already destroyed by
-        // the cublas_handle destructor and we're free to delete the atomic
-        // object.
-        delete ptr;
+        : ih(ih) {
+    if (!device_handle_map) {
+        device_handle_map = std::shared_ptr<std::unordered_map<CUdevice, cublasHandle_t>>(
+            new std::unordered_map<CUdevice, cublasHandle_t>(), [](auto* map) {
+                for (auto& handle_pair : *map) {
+                    cublasStatus_t err;
```

Comment:
Would you need to set a native context here?

Reply:
I don't think so. Why do you think I would have to?

Reply:
Is it valid to destroy the `cuBlasHandle_t` for device 1 when device 0 is the active device (the active native context)?

Reply:
That is a great question

Reply:
From https://docs.nvidia.com/cuda/cublas/index.html?highlight=cublasdestroy#cublasdestroy

> This function releases hardware resources used by the cuBLAS library. This function is usually the last call with a particular handle to the cuBLAS library. Because [cublasCreate()](https://docs.nvidia.com/cuda/cublas/index.html?highlight=cublasdestroy#cublascreate) allocates some internal resources and the release of those resources by calling [cublasDestroy()](https://docs.nvidia.com/cuda/cublas/index.html?highlight=cublasdestroy#cublasdestroy) will implicitly call cudaDeviceSynchronize(), it is recommended to minimize the number of times these functions are called.

This means we don't need to worry about cases where we might destroy the handle while work is still being executed on the device.

Reply:
Yes, but also this will only be called at the end of the program anyway, I mean when the host thread is destroyed. So should be good. 

Regarding setting the native context, I think you're right. I need do need to set it here.

Reply:
Now when I think about it, maybe this code is still too overcomplicated. Do we need the map at all? The current map is `thread_local` but also a single host thread is assigned to max. 1 cuda device. Or is it not?

If a single host thread can have multiple cublas handles tied to different cuda devices then fine, map is necessary. But cublas docs say:

>This approach allows the user to explicitly control the library setup when using multiple host threads and multiple GPUs. For example, the application can use cudaSetDevice() to associate different devices with different host threads and in each of those host threads it can initialize a unique handle to the cuBLAS library context, which will use the particular device associated with that host thread. Then, the cuBLAS library function calls made with different handles will automatically dispatch the computation to different devices.

Which suggests it's just 1 host thread -> 1 cuda device = 1 cublas handle per host thread.